---
title: "Spatial transcriptomics in spaceranger and Seurat workflow"
output: html_notebook
author: "Emily Chambers"
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
---

# Introduction

Aiming to show a complete workthrough of 10X Spatial Gene Expression data from raw reads to analysis and visualization using 10X genomics spaceranger and Seurat.

# Acknowledgements

Tutorial adapted from 10X spaceranger tutorial and Seurat vignette

-   <https://github.com/satijalab/seurat/blob/master/vignettes/spatial_vignette.Rmd>
-   <https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/tutorials/count-ff-tutorial>

# Data

The data is publicly available Mouse Brain Coronal Section

-   Tissue section of 10 Âµm thickness
-   H&E image acquired using a Nikon Ti2-E microscope
-   Sequencing Depth: 115,569 read pairs per spot
-   Sequencing Coverage: Read 1 - 28 bp (includes 16 bp Spatial Barcode, 12 bp UMI); Read 2 - 120 bp (transcript); i7 sample index - 10 bp; i5 sample index - 10 bp
-   Visium Slide: V19L01-041
-   Capture Area: C1

<https://www.10xgenomics.com/resources/datasets/mouse-brain-section-coronal-1-standard-1-1-0>

```{bash}
curl https://s3-us-west-2.amazonaws.com/10x.files/samples/spatial-exp/1.1.0/V1_Adult_Mouse_Brain/V1_Adult_Mouse_Brain_fastqs.tar -o datasets/V1_Adult_Mouse_Brain_fastqs.tar
curl https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Adult_Mouse_Brain/V1_Adult_Mouse_Brain_image.tif -o datasets/V1_Adult_Mouse_Brain_image.tif

# Download mouse reference
curl -O https://cf.10xgenomics.com/supp/spatial-exp/refdata-gex-mm10-2020-A.tar.gz

```

Space ranger was used with the following params:

```{bash}
./spaceranger-2.0.0/spaceranger count --id="V1_Adult_Mouse_Brain" \
                   --description="Adult Mouse Brain (Coronal)" \
                   --transcriptome=refdata-gex-mm10-2020-A \
                   --fastqs=datasets/V1_Adult_Mouse_Brain_fastqs \
                   --image=datasets/V1_Adult_Mouse_Brain_image.tif \
                   --slide=V19L01-041 \
                   --area=C1 \
		               --jobmode=local \
                   --localcores=8 \
		               --localmem=40
		               
```

# Load libraries

```{r, message=FALSE}
library(Seurat)
library(ggplot2)
library(patchwork)
library(dplyr)
```

We are loading in the data directly from the output of spaceranger

Make sure hd5f libraries are installed

```{r}
#brain <- 
Load10X_Spatial(data.dir ="outs")
brain


```

Issue #2

The coordinates of the spatial image are characters instead of integers, this causes issues with the data being read.

```{r}
brain@images[["slice1"]]@coordinates[["tissue"]] <- as.integer(brain@images[["slice1"]]@coordinates[["tissue"]])
brain@images[["slice1"]]@coordinates[["row"]] <- as.integer(brain@images[["slice1"]]@coordinates[["row"]])
brain@images[["slice1"]]@coordinates[["col"]] <- as.integer(brain@images[["slice1"]]@coordinates[["col"]])
brain@images[["slice1"]]@coordinates[["imagerow"]] <- as.integer(brain@images[["slice1"]]@coordinates[["imagerow"]])
brain@images[["slice1"]]@coordinates[["imagecol"]] <- as.integer(brain@images[["slice1"]]@coordinates[["imagecol"]])
```

# Data processing

```{r}
plot1 <- VlnPlot(brain, features = "nCount_Spatial", pt.size = 0.1) + NoLegend()
plot2 <- SpatialFeaturePlot(brain, features = "nCount_Spatial") + theme(legend.position = "right")
wrap_plots(plot1, plot2)

```

```{r}
brain <- SCTransform(brain, assay = "Spatial", verbose = FALSE)
```

```{r}
SpatialFeaturePlot(brain, features = c("Hpca", "Ttr"))
```

```{r}
p1 <- SpatialFeaturePlot(brain, features = "Ttr", pt.size.factor = 1)
p2 <- SpatialFeaturePlot(brain, features = "Ttr", alpha = c(0.1, 1))
p1 + p2
```

```{r}
brain <- RunPCA(brain, assay = "SCT", verbose = FALSE)
brain <- FindNeighbors(brain, reduction = "pca", dims = 1:30)
brain <- FindClusters(brain, verbose = FALSE)
brain <- RunUMAP(brain, reduction = "pca", dims = 1:30)
```

```{r}
p1 <- DimPlot(brain, reduction = "umap", label = TRUE)
p2 <- SpatialDimPlot(brain, label = TRUE, label.size = 3)
p1 + p2
```

```{r}
SpatialDimPlot(brain, cells.highlight = CellsByIdentities(object = brain, idents = c(2, 1, 4, 3,
    5, 8)), facet.highlight = TRUE, ncol = 3)
```

```{r}
SpatialDimPlot(brain, interactive = TRUE) 
```

```{r}
SpatialFeaturePlot(brain, features = "Ttr", interactive = TRUE)
```

# Idenstification of spatially variable features

1st methos : Using prior knowledge

```{r}
de_markers <- FindMarkers(brain, ident.1 = 5, ident.2 = 6)
SpatialFeaturePlot(object = brain, features = rownames(de_markers)[1:3], alpha = c(0.1, 1), ncol = 3)

```

2nd method: Looking for patterns in the data

```{r}
brain <- FindSpatiallyVariableFeatures(brain, assay = "SCT", features = VariableFeatures(brain)[1:1000],
    selection.method = "markvariogram")
top.features <- head(SpatiallyVariableFeatures(brain, selection.method = "markvariogram"), 6)
SpatialFeaturePlot(brain, features = top.features, ncol = 3, alpha = c(0.1, 1))
```
